File: scripts/extract-pricelist-models.py
"""
scripts/extract-pricelist-models.py
Extract {mfr, model, ds} from PRICE_LIST_DATA in dashboard.html -> pricelist-models.json
Run: python scripts/extract-pricelist-models.py
"""
import re, json, os, sys

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
HTML_PATH = os.path.join(ROOT, 'dashboard.html')
OUT_PATH  = os.path.join(ROOT, 'pricelist-models.json')

with open(HTML_PATH, encoding='utf-8') as f:
    html = f.read()

# ── Locate the PRICE_LIST_DATA array ───────────────────────────────────────────
marker = 'const PRICE_LIST_DATA'  # matches both "= [" and "=["
start  = html.find(marker)
if start == -1:
    print('ERROR: PRICE_LIST_DATA not found'); sys.exit(1)

arr_start = html.index('[', start)

# Bracket-match to find the end
depth = 0
arr_end = -1
for i in range(arr_start, len(html)):
    c = html[i]
    if   c == '[': depth += 1
    elif c == ']':
        depth -= 1
        if depth == 0: arr_end = i; break
if arr_end == -1:
    print('ERROR: could not find closing ] for PRICE_LIST_DATA'); sys.exit(1)

array_text = html[arr_start:arr_end+1]

# ── Extract each object ──────────────────────────────────────────────────────
# Regex for one field: key: 'value'  or  key: "value"
def grab(text, key):
    m = re.search(r'\b' + key + r'\s*:\s*[\'"]([^\'\"]*)[\'"]', text)
    return m.group(1) if m else ''

# Split on object boundaries: find each { ... } at depth-1 of the array
items = []
seen  = set()
obj_re = re.compile(r'\{[^{}]*\}', re.DOTALL)
for m in obj_re.finditer(array_text):
    obj = m.group(0)
    mfr   = grab(obj, 'mfr')
    model = grab(obj, 'model')
    ds    = grab(obj, 'ds')
    if not mfr or not model:
        continue
    key = f'{mfr}||{model}'
    if key in seen:
        continue
    seen.add(key)
    items.append({'mfr': mfr, 'model': model, 'ds': ds})

with open(OUT_PATH, 'w', encoding='utf-8') as f:
    json.dump(items, f, ensure_ascii=False, indent=2)

print(f'Extracted {len(items)} unique (mfr, model) entries -> pricelist-models.json')
by_mfr = {}
for x in items:
    by_mfr[x['mfr']] = by_mfr.get(x['mfr'], 0) + 1
for mfr, n in sorted(by_mfr.items(), key=lambda t: -t[1]):
    has_ds = sum(1 for x in items if x['mfr']==mfr and x['ds'])
    print(f"  {mfr:<14} {n:>4} items  ({has_ds} with existing ds url)")

File: scripts/extract-pricelist-models.js
/**
 * scripts/extract-pricelist-models.js
 *
 * One-time (and on-demand) script that extracts {mfr, model, ds} entries
 * from the PRICE_LIST_DATA constant in dashboard.html and writes them to
 * pricelist-models.json in the project root.
 *
 * Run: node scripts/extract-pricelist-models.js
 * Re-run any time you edit PRICE_LIST_DATA in dashboard.html.
 */

'use strict';
const fs   = require('fs');
const path = require('path');
const vm   = require('vm');

const rootDir   = path.join(__dirname, '..');
const htmlPath  = path.join(rootDir, 'dashboard.html');
const outPath   = path.join(rootDir, 'pricelist-models.json');

const html = fs.readFileSync(htmlPath, 'utf8');

// Locate the start of the array literal
const MARKER = 'const PRICE_LIST_DATA = ';
const markerIdx = html.indexOf(MARKER);
if (markerIdx === -1) {
  console.error('ERROR: PRICE_LIST_DATA constant not found in dashboard.html');
  process.exit(1);
}

const arrayStart = html.indexOf('[', markerIdx);
if (arrayStart === -1) {
  console.error('ERROR: could not find opening [ for PRICE_LIST_DATA');
  process.exit(1);
}

// Bracket-match to find the end of the array
let depth = 0;
let arrayEnd = -1;
for (let i = arrayStart; i < html.length; i++) {
  const c = html[i];
  if      (c === '[') depth++;
  else if (c === ']') { depth--; if (depth === 0) { arrayEnd = i; break; } }
}
if (arrayEnd === -1) {
  console.error('ERROR: could not find closing ] for PRICE_LIST_DATA');
  process.exit(1);
}

const arrayStr = html.slice(arrayStart, arrayEnd + 1);

// Evaluate safely in a sandboxed context
const sandbox = {};
try {
  vm.runInNewContext(`data = ${arrayStr}`, sandbox);
} catch (e) {
  console.error('ERROR evaluating PRICE_LIST_DATA:', e.message);
  process.exit(1);
}

const raw = sandbox.data;
if (!Array.isArray(raw)) {
  console.error('ERROR: evaluated value is not an array');
  process.exit(1);
}

// Extract only the fields the ds-finder needs
const models = raw
  .map(item => ({
    mfr:   (item.mfr   || '').trim(),
    model: (item.model || '').trim(),
    ds:    (item.ds    || '').trim(),
  }))
  .filter(x => x.mfr && x.model);

// Deduplicate by (mfr, model)
const seen = new Set();
const unique = models.filter(x => {
  const key = `${x.mfr}||${x.model}`;
  if (seen.has(key)) return false;
  seen.add(key);
  return true;
});

fs.writeFileSync(outPath, JSON.stringify(unique, null, 2), 'utf8');
console.log(`✅ Extracted ${unique.length} unique (mfr, model) entries → pricelist-models.json`);

// Print a quick breakdown by manufacturer
const byMfr = {};
for (const x of unique) { byMfr[x.mfr] = (byMfr[x.mfr] || 0) + 1; }
for (const [mfr, n] of Object.entries(byMfr).sort((a,b) => b[1]-a[1])) {
  const hasDsCount = unique.filter(x => x.mfr === mfr && x.ds).length;
  console.log(`  ${mfr.padEnd(12)} ${String(n).padStart(4)} items  (${hasDsCount} with existing ds url)`);
}
